pxtservices/iframeDriver.ts

```diff
@@ -0,0 +1,553 @@
+                action: "requestprojectcloudstatus",
+                headerIds
+            } as pxt.editor.EditorMessageRequestProjectCloudStatus
+        );
+    }
+
+    addEventListener(event: typeof MessageSentEvent, handler: (ev: pxt.editor.EditorMessage) => void): void;
+    addEventListener(event: typeof MessageReceivedEvent, handler: (ev: pxt.editor.EditorMessage) => void): void;
+    addEventListener(event: "event", handler: (ev: pxt.editor.EditorMessageEventRequest) => void): void;
+    addEventListener(event: "simevent", handler: (ev: pxt.editor.EditorSimulatorEvent) => void): void;
+    addEventListener(event: "tutorialevent", handler: (ev: pxt.editor.EditorMessageTutorialEventRequest) => void): void;
+    addEventListener(event: "editorcontentloaded", handler: (ev: pxt.editor.EditorContentLoadedRequest) => void): void;
+    addEventListener(event: "workspacesave", handler: (ev: pxt.editor.EditorWorkspaceSaveRequest) => void): void;
+    addEventListener(event: "workspaceevent", handler: (ev: pxt.editor.EditorWorkspaceEvent) => void): void;
+    addEventListener(event: "workspacereset", handler: (ev: pxt.editor.EditorWorkspaceSyncRequest) => void): void;
+    addEventListener(event: "workspacesync", handler: (ev: pxt.editor.EditorWorkspaceSyncRequest) => void): void;
+    addEventListener(event: "workspaceloaded", handler: (ev: pxt.editor.EditorWorkspaceSyncRequest) => void): void;
+    addEventListener(event: "workspacediagnostics", handler: (ev: pxt.editor.EditorWorkspaceDiagnostics) => void): void;
+    addEventListener(event: "editorcontentloaded", handler: (ev: pxt.editor.EditorContentLoadedRequest) => void): void;
+    addEventListener(event: "projectcloudstatus", handler: (ev: pxt.editor.EditorMessageProjectCloudStatus) => void): void;
+    addEventListener(event: string, handler: (ev: any) => void): void {
+        if (!this.editorEventListeners[event]) this.editorEventListeners[event] = [];
+        this.editorEventListeners[event].push(handler);
+    }
+
+    removeEventListener(event: string, handler: (ev: any) => void) {
+        if (this.editorEventListeners[event]) {
+            const filtered = this.editorEventListeners[event].filter(h => h !== handler);
+
+            if (filtered.length === 0) {
+                delete this.editorEventListeners[event];
+            }
+            else {
+                this.editorEventListeners[event] = filtered;
+            }
+        }
+    }
+
+    sendMessage(message: pxt.editor.EditorMessageRequest): Promise<pxt.editor.EditorMessageResponse> {
+        return this.sendRequest(message) as Promise<pxt.editor.EditorMessageResponse>;
+    }
+
+    protected onMessageReceived = (event: MessageEvent) => {
+        const data = event.data as pxt.editor.EditorMessageRequest;
+        if (!data || !/^pxt(host|editor|pkgext|sim)$/.test(data.type)) return;
+
+        if (data.type === "pxteditor") {
+            if (data.id && this.pendingMessages[data.id]) {
```

Comment:
Noting a potential leak: The promise associated with sent messages that never receive a their expected reply could hang forever. This driver may want to run a pending message checker in the background (on a timeout), to clear pending messages that haven't received a reply in a certain amount of time.